docarray/typing/tensor/tensor.py

```diff
@@ -1,22 +1,145 @@
 
 torch_available = is_torch_available()
 if torch_available:
-    from docarray.typing.tensor.torch_tensor import TorchTensor  # noqa: F401
+    import torch
 
+    from docarray.typing.tensor.torch_tensor import TorchTensor  # noqa: F401
 
 tf_available = is_tf_available()
 if tf_available:
+    import tensorflow as tf  # type: ignore
+
     from docarray.typing.tensor.tensorflow_tensor import TensorFlowTensor  # noqa: F401
 
 
-AnyTensor = Union[NdArray]
-if torch_available and tf_available:
-    AnyTensor = Union[NdArray, TorchTensor, TensorFlowTensor]  # type: ignore
-elif torch_available:
-    AnyTensor = Union[NdArray, TorchTensor]  # type: ignore
-elif tf_available:
-    AnyTensor = Union[NdArray, TensorFlowTensor]  # type: ignore
+if TYPE_CHECKING:
+    from pydantic import BaseConfig
+    from pydantic.fields import ModelField
+
+    # Below is the hack to make the type checker happy. But `AnyTensor` is defined as a class and with same underlying
+    # behavior as `Union[TorchTensor, TensorFlowTensor, NdArray]` so it should be fine to use `AnyTensor` as
+    # the type for `tensor` field in `BaseDoc` class.
+    AnyTensor = Union[NdArray]
+    if torch_available and tf_available:
+        AnyTensor = Union[NdArray, TorchTensor, TensorFlowTensor]  # type: ignore
+    elif torch_available:
+        AnyTensor = Union[NdArray, TorchTensor]  # type: ignore
+    elif tf_available:
+        AnyTensor = Union[NdArray, TensorFlowTensor]  # type: ignore
+
+else:
+
+    T = TypeVar("T", bound="AnyTensor")
+    ShapeT = TypeVar('ShapeT')
+
+    class AnyTensor(AbstractTensor, Generic[ShapeT]):
+        """
+        Represents a tensor object that can be used with TensorFlow, PyTorch, and NumPy type.
+
+        ---
+
+        '''python
+        from docarray import BaseDoc
+        from docarray.typing import AnyTensor
+
+
+        class MyTensorDoc(BaseDoc):
+            tensor: AnyTensor
+
+
+        # Example usage with TensorFlow:
+        import tensorflow as tf
+
+        doc = MyTensorDoc(tensor=tf.zeros(1000, 2))
+
+        # Example usage with PyTorch:
+        import torch
+
+        doc = MyTensorDoc(tensor=torch.zeros(1000, 2))
+
+        # Example usage with NumPy:
+        import numpy as np
+
+        doc = MyTensorDoc(tensor=np.zeros((1000, 2)))
+        '''
+
+        Returns:
+            Union[TorchTensor, TensorFlowTensor, NdArray]: The validated and converted tensor.
+
+        Raises:
+            TypeError: If the input value is not a compatible type (torch.Tensor, tensorflow.Tensor, numpy.ndarray).
+
+        """
+
+        def __getitem__(self: T, item):
+            pass
+
+        def __setitem__(self, index, value):
+            pass
+
+        def __iter__(self):
+            pass
+
+        def __len__(self):
+            pass
+
+        @classmethod
+        def _docarray_from_native(cls: Type[T], value: Any):
+            raise AttributeError(f'This method should not be called on {cls}.')
```

Comment:
Maybe RuntimeError is more clear, Technically this method is nto an Attribute right ?

Reply:
Technically, [methods are attributes](https://docs.python.org/3/tutorial/classes.html#:~:text=By%20the%20way%2C%20I%20use%20the%20word%20attribute%20for%20any%20name%20following%20a%20dot) of class instance but you are right, this error will pop up during runtime so raising it as `RuntimeError` should be more apt